sql/pgwire/v3.go

```diff
@@ -0,0 +1,284 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Ben Darnell
+
+package pgwire
+
+import (
+	"bufio"
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"net"
+
+	"github.com/cockroachdb/cockroach/sql"
+	"github.com/cockroachdb/cockroach/sql/driver"
+	"github.com/cockroachdb/cockroach/util"
+	"github.com/cockroachdb/cockroach/util/log"
+	"github.com/lib/pq/oid"
+)
+
+type messageType byte
+
+const (
+	serverMsgAuth            messageType = 'R'
+	serverMsgCommandComplete             = 'C'
+	serverMsgDataRow                     = 'D'
+	serverMsgErrorResponse               = 'E'
+	serverMsgParseComplete               = '1'
+	serverMsgReady                       = 'Z'
+	serverMsgRowDescription              = 'T'
+
+	clientMsgSimpleQuery = 'Q'
+	clientMsgParse       = 'P'
+)
+
+const (
+	authOK int32 = 0
+)
+
+type parsedQuery struct {
+	query string
+	types []oid.Oid
+}
+
+type v3Conn struct {
+	conn     net.Conn
+	opts     map[string]string
+	executor sql.Executor
+	parsed   map[string]parsedQuery
+}
+
+func newV3Conn(conn net.Conn, data []byte, executor sql.Executor) (*v3Conn, error) {
+	v3conn := &v3Conn{
+		conn:     conn,
+		opts:     map[string]string{},
+		executor: executor,
+		parsed:   map[string]parsedQuery{},
+	}
+	if err := v3conn.parseOptions(data); err != nil {
+		return nil, err
+	}
+	return v3conn, nil
+}
+
+func (c *v3Conn) parseOptions(data []byte) error {
+	buf := bytes.NewBuffer(data)
+	for {
+		key, err := readString(buf)
+		if err != nil {
+			return util.Errorf("error reading option key: %s", err)
+		}
+		if len(key) == 0 {
+			return nil
+		}
+		value, err := readString(buf)
+		if err != nil {
+			return util.Errorf("error reading option value: %s", err)
+		}
+		c.opts[key] = value
+	}
+}
+
+func (c *v3Conn) serve() error {
+	rd := bufio.NewReader(c.conn)
+	// TODO(bdarnell): real auth flow. For now just accept anything.
+	var authOKMsg bytes.Buffer
+	if err := binary.Write(&authOKMsg, binary.BigEndian, authOK); err != nil {
+		return err
+	}
+	if err := writeTypedMsg(c.conn, serverMsgAuth, authOKMsg.Bytes()); err != nil {
+		return err
+	}
+	for {
+		// TODO(bdarnell): change the 'I' below based on transaction status.
+		if err := writeTypedMsg(c.conn, serverMsgReady, []byte{'I'}); err != nil {
+			return err
+		}
+		typ, msg, err := readTypedMsg(rd)
+		if err != nil {
+			return err
+		}
+		buf := bytes.NewBuffer(msg)
+		switch typ {
+		case clientMsgSimpleQuery:
+			err = c.handleSimpleQuery(buf)
+
+		case clientMsgParse:
+			err = c.handleParse(buf)
+
+		default:
+			log.Fatalf("unrecognized client message type %c", typ)
```

Comment:
Fatal doesn't panic, I believe it just calls os.Exit(something).
On Sep 14, 2015 2:01 PM, "mrtracy" <notifications@github.com> wrote:

> In sql/pgwire/v3.go
> <https://github.com/cockroachdb/cockroach/pull/2491#discussion_r39425437>:
>
> > +			return err
> > +		}
> > +		typ, msg, err := readTypedMsg(rd)
> > +		if err != nil {
> > +			return err
> > +		}
> > +		buf := bytes.NewBuffer(msg)
> > +		switch typ {
> > +		case clientMsgSimpleQuery:
> > +			err = c.handleSimpleQuery(buf)
> > +
> > +		case clientMsgParse:
> > +			err = c.handleParse(buf)
> > +
> > +		default:
> > +			log.Fatalf("unrecognized client message type %c", typ)
>
> Should this be a Fatal? Seems that it would crash the server based on a
> malformed client message, unless the resulting panic is being caught
> somewhere.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/2491/files#r39425437>.
>